fem/kernels.hpp

```diff
@@ -51,192 +95,996 @@ MFEM_HOST_DEVICE inline void LoadBG(const int D1D, const int Q1D,
    MFEM_SYNC_THREAD;
 }
 
+/// Load Bt1d & Gt1d matrices into shared memory
+template<int MD1, int MQ1>
+MFEM_HOST_DEVICE inline void LoadBGt(const int D1D, const int Q1D,
+                                     const ConstDeviceMatrix &b,
+                                     const ConstDeviceMatrix &g,
+                                     double (&sBG)[2][MQ1*MD1])
+{
+   const int tidz = MFEM_THREAD_ID(z);
+   DeviceMatrix Bt(sBG[0], MQ1, MD1);
+   DeviceMatrix Gt(sBG[1], MQ1, MD1);
+
+   if (tidz == 0)
+   {
+      MFEM_FOREACH_THREAD(d,y,D1D)
+      {
+         MFEM_FOREACH_THREAD(q,x,Q1D)
+         {
+            Bt(q,d) = b(q,d);
+            Gt(q,d) = g(q,d);
+         }
+      }
+   }
+   MFEM_SYNC_THREAD;
+}
+
+/// Load 2D input scalar into shared memory
+template<int MD1, int NBZ>
+MFEM_HOST_DEVICE inline void LoadX(const int e, const int D1D,
+                                   const DeviceTensor<3, const double> &x,
+                                   double (&sX)[NBZ][MD1*MD1])
+{
+   const int tidz = MFEM_THREAD_ID(z);
+   DeviceMatrix X(sX[tidz], MD1, MD1);
```

Comment:
Is it `MD1` to optimize shared memory conflicts?
```suggestion
   DeviceMatrix X(sX[tidz], D1D, D1D);
```

Reply:
`done`